File: api/scheduler.py
import asyncio
import logging

from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger

from db.config import settings
from mediafusion_scrapy.task import run_spider
from scrapers.background_scraper import run_background_search
from scrapers.dmm_hashlist import run_dmm_hashlist_scraper
from scrapers.feed_scraper import run_jackett_feed_scraper, run_prowlarr_feed_scraper
from scrapers.non_torrent_background_scraper import (
    run_acestream_background_scraper,
    run_telegram_background_scraper,
    run_youtube_background_scraper,
)
from scrapers.rss_scraper import run_rss_feed_scraper
from scrapers.scraper_tasks import cleanup_expired_scraper_task
from scrapers.trackers import update_torrent_seeders
from scrapers.tv import validate_tv_streams_in_db
from api.services.sync.tasks import run_all_integration_syncs
from streaming_providers.cache_helpers import cleanup_expired_cache
from utils.telegram_bot import telegram_notifier

logger = logging.getLogger(__name__)


async def async_send(actor_send_method, **kwargs):
    """
    Wrapper to enqueue background jobs with retry.
    """
    max_attempts = 3
    delay_seconds = 1
    actor_name = getattr(actor_send_method, "__qualname__", repr(actor_send_method))

    for attempt in range(1, max_attempts + 1):
        try:
            await actor_send_method(**kwargs)
            return
        except Exception as exc:
            if attempt >= max_attempts:
                logger.warning(
                    "Failed to enqueue scheduler actor %s after %d attempts: %s",
                    actor_name,
                    max_attempts,
                    exc,
                )
                return
            await asyncio.sleep(delay_seconds)
            delay_seconds *= 2


def setup_scheduler(scheduler: AsyncIOScheduler):
    """
    Set up the scheduler with the required jobs.
    All jobs use async_send wrapper with retry for Redis transient failures.
    """
    # Setup tamil blasters scraper
    if not settings.disable_tamil_blasters_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.tamil_blasters_scheduler_crontab),
            name="tamil_blasters",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "tamil_blasters",
                "crontab_expression": settings.tamil_blasters_scheduler_crontab,
            },
        )

    # Setup tamilmv scraper
    if not settings.disable_tamilmv_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.tamilmv_scheduler_crontab),
            name="tamilmv",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "tamilmv",
                "crontab_expression": settings.tamilmv_scheduler_crontab,
            },
        )

    # Setup formula_ext scraper
    if not settings.disable_formula_ext_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.formula_ext_scheduler_crontab),
            name="formula_ext",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "formula_ext",
                "scrape_all": "false",
                "crontab_expression": settings.formula_ext_scheduler_crontab,
            },
        )

    # Setup motogp_ext scraper
    if not settings.disable_motogp_ext_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.motogp_ext_scheduler_crontab),
            name="motogp_ext",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "motogp_ext",
                "scrape_all": "false",
                "crontab_expression": settings.motogp_ext_scheduler_crontab,
            },
        )

    # Setup wwe_ext scraper
    if not settings.disable_wwe_ext_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.wwe_ext_scheduler_crontab),
            name="wwe_ext",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "wwe_ext",
                "scrape_all": "false",
                "crontab_expression": settings.wwe_ext_scheduler_crontab,
            },
        )

    # Setup ufc_ext scraper
    if not settings.disable_ufc_ext_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.ufc_ext_scheduler_crontab),
            name="ufc_ext",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "ufc_ext",
                "scrape_all": "false",
                "crontab_expression": settings.ufc_ext_scheduler_crontab,
            },
        )

    # Setup movies_tv_ext scraper
    if not settings.disable_movies_tv_ext_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.movies_tv_ext_scheduler_crontab),
            name="movies_tv_ext",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "movies_tv_ext",
                "scrape_all": "false",
                "crontab_expression": settings.movies_tv_ext_scheduler_crontab,
            },
        )

    # Setup nowmetv scraper
    if not settings.disable_nowmetv_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.nowmetv_scheduler_crontab),
            name="nowmetv",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "nowmetv",
                "crontab_expression": settings.nowmetv_scheduler_crontab,
            },
        )

    # Setup nowsports scraper
    if not settings.disable_nowsports_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.nowsports_scheduler_crontab),
            name="nowsports",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "nowsports",
                "crontab_expression": settings.nowsports_scheduler_crontab,
            },
        )

    # Setup tamilultra scraper
    if not settings.disable_tamilultra_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.tamilultra_scheduler_crontab),
            name="tamilultra",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "tamilultra",
                "crontab_expression": settings.tamilultra_scheduler_crontab,
            },
        )

    # Schedule validate_tv_streams_in_db
    if not settings.disable_validate_tv_streams_in_db:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.validate_tv_streams_in_db_crontab),
            name="validate_tv_streams_in_db",
            kwargs={
                "actor_send_method": validate_tv_streams_in_db.async_send,
                "crontab_expression": settings.validate_tv_streams_in_db_crontab,
            },
        )

    # Schedule sport_video scraper
    if not settings.disable_sport_video_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.sport_video_scheduler_crontab),
            name="sport_video",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "sport_video",
                "scrape_all": "false",
                "crontab_expression": settings.sport_video_scheduler_crontab,
            },
        )

    # Schedule dlhd scraper
    if not settings.disable_dlhd_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.dlhd_scheduler_crontab),
            name="dlhd",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "dlhd",
                "crontab_expression": settings.dlhd_scheduler_crontab,
            },
        )

    if not settings.disable_update_seeders:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.update_seeders_crontab),
            name="update_seeders",
            kwargs={
                "actor_send_method": update_torrent_seeders.async_send,
                "crontab_expression": settings.update_seeders_crontab,
            },
        )

    if not settings.disable_arab_torrents_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.arab_torrents_scheduler_crontab),
            name="arab_torrents",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "arab_torrents",
                "crontab_expression": settings.arab_torrents_scheduler_crontab,
            },
        )

    if not settings.disable_x1337_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.x1337_scheduler_crontab),
            name="x1337",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "x1337",
                "scrape_all": "false",
                "crontab_expression": settings.x1337_scheduler_crontab,
            },
        )

    if not settings.disable_thepiratebay_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.thepiratebay_scheduler_crontab),
            name="thepiratebay",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "thepiratebay",
                "scrape_all": "false",
                "crontab_expression": settings.thepiratebay_scheduler_crontab,
            },
        )

    if not settings.disable_rutor_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.rutor_scheduler_crontab),
            name="rutor",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "rutor",
                "scrape_all": "false",
                "crontab_expression": settings.rutor_scheduler_crontab,
            },
        )

    if not settings.disable_limetorrents_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.limetorrents_scheduler_crontab),
            name="limetorrents",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "limetorrents",
                "scrape_all": "false",
                "crontab_expression": settings.limetorrents_scheduler_crontab,
            },
        )

    if not settings.disable_yts_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.yts_scheduler_crontab),
            name="yts",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "yts",
                "scrape_all": "false",
                "crontab_expression": settings.yts_scheduler_crontab,
            },
        )

    if not settings.disable_bt4g_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.bt4g_scheduler_crontab),
            name="bt4g",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "bt4g",
                "scrape_all": "false",
                "crontab_expression": settings.bt4g_scheduler_crontab,
            },
        )

    if not settings.disable_nyaa_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.nyaa_scheduler_crontab),
            name="nyaa",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "nyaa",
                "scrape_all": "false",
                "crontab_expression": settings.nyaa_scheduler_crontab,
            },
        )

    if not settings.disable_animetosho_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.animetosho_scheduler_crontab),
            name="animetosho",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "animetosho",
                "scrape_all": "false",
                "crontab_expression": settings.animetosho_scheduler_crontab,
            },
        )

    if not settings.disable_subsplease_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.subsplease_scheduler_crontab),
            name="subsplease",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "subsplease",
                "scrape_all": "false",
                "crontab_expression": settings.subsplease_scheduler_crontab,
            },
        )

    if not settings.disable_animepahe_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.animepahe_scheduler_crontab),
            name="animepahe",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "animepahe",
                "scrape_all": "false",
                "crontab_expression": settings.animepahe_scheduler_crontab,
            },
        )

    if not settings.disable_bt52_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.bt52_scheduler_crontab),
            name="bt52",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "bt52",
                "scrape_all": "false",
                "crontab_expression": settings.bt52_scheduler_crontab,
            },
        )

    if not settings.disable_uindex_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.uindex_scheduler_crontab),
            name="uindex",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "uindex",
                "scrape_all": "false",
                "crontab_expression": settings.uindex_scheduler_crontab,
            },
        )

    if not settings.disable_eztv_rss_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.eztv_rss_scheduler_crontab),
            name="eztv_rss",
            kwargs={
                "actor_send_method": run_spider.async_send,
                "spider_name": "eztv_rss",
                "crontab_expression": settings.eztv_rss_scheduler_crontab,
            },
        )

    # Schedule the feed scraper
    if not settings.disable_prowlarr_feed_scraper:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.prowlarr_feed_scraper_crontab),
            name="prowlarr_feed_scraper",
            kwargs={
                "actor_send_method": run_prowlarr_feed_scraper.async_send,
                "crontab_expression": settings.prowlarr_feed_scraper_crontab,
            },
        )

    if not settings.disable_jackett_feed_scraper:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.jackett_feed_scraper_crontab),
            name="jackett_feed_scraper",
            kwargs={
                "actor_send_method": run_jackett_feed_scraper.async_send,
                "crontab_expression": settings.jackett_feed_scraper_crontab,
            },
        )

    # Schedule RSS feed scraper
    if not settings.disable_rss_feed_scraper:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.rss_feed_scraper_crontab),
            name="rss_feed_scraper",
            kwargs={
                "actor_send_method": run_rss_feed_scraper.async_send,
                "crontab_expression": settings.rss_feed_scraper_crontab,
            },
        )

    # Schedule DMM hashlist scraper
    if not settings.disable_dmm_hashlist_scraper:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.dmm_hashlist_scraper_crontab),
            name="dmm_hashlist_scraper",
            kwargs={
                "actor_send_method": run_dmm_hashlist_scraper.async_send,
                "crontab_expression": settings.dmm_hashlist_scraper_crontab,
            },
        )

    if not settings.disable_youtube_background_scraper and settings.is_scrap_from_youtube_background:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.youtube_background_scraper_crontab),
            name="youtube_background_scraper",
            kwargs={
                "actor_send_method": run_youtube_background_scraper.async_send,
                "crontab_expression": settings.youtube_background_scraper_crontab,
            },
        )

    if not settings.disable_acestream_background_scraper and settings.is_scrap_from_acestream_background:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.acestream_background_scraper_crontab),
            name="acestream_background_scraper",
            kwargs={
                "actor_send_method": run_acestream_background_scraper.async_send,
                "crontab_expression": settings.acestream_background_scraper_crontab,
            },
        )

    if not settings.disable_telegram_background_scraper and settings.is_scrap_from_telegram_background:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.telegram_background_scraper_crontab),
            name="telegram_background_scraper",
            kwargs={
                "actor_send_method": run_telegram_background_scraper.async_send,
                "crontab_expression": settings.telegram_background_scraper_crontab,
            },
        )

    scheduler.add_job(
        async_send,
        CronTrigger.from_crontab(settings.cleanup_expired_scraper_task_crontab),
        name="cleanup_expired_scraper_task",
        kwargs={
            "actor_send_method": cleanup_expired_scraper_task.async_send,
            "crontab_expression": settings.cleanup_expired_scraper_task_crontab,
        },
    )

    scheduler.add_job(
        async_send,
        CronTrigger.from_crontab(settings.cleanup_expired_cache_task_crontab),
        name="cleanup_expired_cache_task",
        kwargs={
            "actor_send_method": cleanup_expired_cache.async_send,
            "crontab_expression": settings.cleanup_expired_cache_task_crontab,
        },
    )

    scheduler.add_job(
        async_send,
        CronTrigger.from_crontab(settings.background_search_crontab),
        name="background_search",
        kwargs={
            "actor_send_method": run_background_search.async_send,
            "crontab_expression": settings.background_search_crontab,
        },
    )

    if (
        not settings.disable_pending_moderation_reminder_scheduler
        and settings.telegram_bot_token
        and settings.telegram_chat_id
    ):
        scheduler.add_job(
            telegram_notifier.send_pending_moderation_reminder,
            CronTrigger.from_crontab(settings.pending_moderation_reminder_crontab),
            name="pending_moderation_reminder",
        )

    if not settings.disable_integration_sync_scheduler:
        scheduler.add_job(
            async_send,
            CronTrigger.from_crontab(settings.integration_sync_crontab),
            name="integration_sync",
            kwargs={
                "actor_send_method": run_all_integration_syncs.async_send,
            },
        )

File: services/api/src/scheduler.rs
use std::str::FromStr;
/// Background scheduler — mirrors workers/scheduler.py.
///
/// Each enabled job spawns a dedicated tokio task that:
///   1. Parses its cron expression at startup.
///   2. Sleeps until the next scheduled tick.
///   3. Enqueues the corresponding taskiq actor via Redis Streams.
///   4. Loops back to step 2.
///
/// The taskiq wire format (XADD + task-record SET) is identical to the
/// `enqueue_taskiq` helper already used in admin_scrapers.rs.
use std::sync::Arc;

use chrono::Utc;
use cron::Schedule;
use fred::prelude::{KeysInterface, ListInterface, StreamsInterface};
use serde_json::{json, Value};
use tracing::{debug, info, warn};

use crate::state::AppState;

// ── Wire format ───────────────────────────────────────────────────────────────

async fn enqueue(state: &Arc<AppState>, actor_name: &str, queue: &str, kwargs: Value) {
    let task_id = uuid::Uuid::new_v4().simple().to_string();
    let queue_key = format!("mediafusion:taskiq:{queue}");

    let mut kw_map = match kwargs {
        Value::Object(m) => m,
        _ => serde_json::Map::new(),
    };
    kw_map.insert("_taskiq_task_id".into(), json!(task_id));

    let msg = json!({
        "task_id": task_id,
        "task_name": actor_name,
        "labels": { "queue_name": queue_key },
        "labels_types": null,
        "args": [],
        "kwargs": Value::Object(kw_map.clone()),
    });
    let msg_json = match serde_json::to_string(&msg) {
        Ok(s) => s,
        Err(e) => {
            warn!("scheduler: serialize {actor_name}: {e}");
            return;
        }
    };

    if let Err(e) = state
        .redis
        .xadd::<String, _, _, _, _>(
            &queue_key,
            false,
            None,
            "*",
            vec![("data", msg_json.as_str())],
        )
        .await
    {
        warn!("scheduler: xadd {actor_name}: {e}");
        return;
    }

    // Remove the internal marker from the stored record
    kw_map.remove("_taskiq_task_id");
    let record = json!({
        "task_id": task_id,
        "actor_name": actor_name,
        "queue_name": queue_key,
        "args_payload": [],
        "kwargs_payload": Value::Object(kw_map),
        "status": "pending",
        "created_at": Utc::now().to_rfc3339(),
    });
    let record_json = match serde_json::to_string(&record) {
        Ok(s) => s,
        Err(e) => {
            warn!("scheduler: serialize record {actor_name}: {e}");
            return;
        }
    };
    let task_key = format!("mediafusion:taskiq:task:{task_id}");
    if let Err(e) = state
        .redis
        .set::<String, _, _>(
            &task_key,
            record_json.as_str(),
            Some(fred::types::Expiration::EX(7 * 24 * 3600)),
            None,
            false,
        )
        .await
    {
        warn!("scheduler: set record {actor_name}: {e}");
    }
    let _: Result<i64, _> = state
        .redis
        .lpush("mediafusion:taskiq:tasks:recent", task_id.as_str())
        .await;
    let _: Result<(), _> = state
        .redis
        .ltrim("mediafusion:taskiq:tasks:recent", 0, 1999)
        .await;

    debug!("scheduler: enqueued {actor_name} → {queue} (task_id={task_id})");
}

// ── Cron expression conversion ────────────────────────────────────────────────
//
// Python/Unix cron uses 5 fields: min hour dom month dow
// The `cron` crate v0.12 requires 7 fields:  sec min hour dom month dow year
// Convert by prepending "0" (fire at :00 seconds) and appending "*" (any year).

fn to_cron_schedule(five_field: &str) -> Result<Schedule, cron::error::Error> {
    let seven_field = format!("0 {} *", five_field);
    Schedule::from_str(&seven_field)
}

// ── Per-job spawn helper ──────────────────────────────────────────────────────

fn spawn_job<F, Fut>(name: &'static str, crontab: String, state: Arc<AppState>, f: F)
where
    F: Fn(Arc<AppState>) -> Fut + Send + 'static,
    Fut: std::future::Future<Output = ()> + Send + 'static,
{
    let schedule = match to_cron_schedule(&crontab) {
        Ok(s) => s,
        Err(e) => {
            warn!("scheduler: invalid crontab for {name} ({crontab:?}): {e}");
            return;
        }
    };

    tokio::spawn(async move {
        info!("scheduler: {name} registered with crontab {crontab}");
        loop {
            let now = Utc::now();
            let Some(next) = schedule.upcoming(Utc).next() else {
                break;
            };
            let delay = (next - now).to_std().unwrap_or_default();
            tokio::time::sleep(delay).await;
            debug!("scheduler: firing {name}");
            f(Arc::clone(&state)).await;
        }
    });
}

// ── Spider helper ─────────────────────────────────────────────────────────────

async fn enqueue_spider(
    state: Arc<AppState>,
    spider_name: &'static str,
    crontab: &str,
    scrape_all: Option<&'static str>,
) {
    let mut kw = serde_json::Map::new();
    kw.insert("spider_name".into(), json!(spider_name));
    kw.insert("crontab_expression".into(), json!(crontab));
    if let Some(v) = scrape_all {
        kw.insert("scrape_all".into(), json!(v));
    }
    enqueue(&state, "run_spider", "scrapy", Value::Object(kw)).await;
}

// ── Public entry point ────────────────────────────────────────────────────────

/// Spawn all enabled scheduler jobs. Called once from `main` after AppState is built.
pub fn start(state: Arc<AppState>) {
    let cfg = &state.config;

    if cfg.disable_all_scheduler {
        info!("scheduler: DISABLE_ALL_SCHEDULER=true — all jobs suppressed");
        return;
    }

    // ── Scrapy spiders ────────────────────────────────────────────────────────

    macro_rules! spider {
        ($name:literal, $crontab:expr, $disabled:expr) => {
            spider!($name, $crontab, $disabled, None)
        };
        ($name:literal, $crontab:expr, $disabled:expr, $scrape_all:expr) => {
            if !$disabled {
                let crontab = $crontab.clone();
                let crontab2 = crontab.clone();
                spawn_job($name, crontab, Arc::clone(&state), move |s| {
                    let ct = crontab2.clone();
                    async move { enqueue_spider(s, $name, &ct, $scrape_all).await }
                });
            }
        };
    }

    spider!(
        "tamilmv",
        cfg.tamilmv_scheduler_crontab,
        cfg.disable_tamilmv_scheduler
    );
    spider!(
        "tamil_blasters",
        cfg.tamil_blasters_scheduler_crontab,
        cfg.disable_tamil_blasters_scheduler
    );
    spider!(
        "formula_ext",
        cfg.formula_ext_scheduler_crontab,
        cfg.disable_formula_ext_scheduler,
        Some("false")
    );
    spider!(
        "motogp_ext",
        cfg.motogp_ext_scheduler_crontab,
        cfg.disable_motogp_ext_scheduler,
        Some("false")
    );
    spider!(
        "wwe_ext",
        cfg.wwe_ext_scheduler_crontab,
        cfg.disable_wwe_ext_scheduler,
        Some("false")
    );
    spider!(
        "ufc_ext",
        cfg.ufc_ext_scheduler_crontab,
        cfg.disable_ufc_ext_scheduler,
        Some("false")
    );
    spider!(
        "movies_tv_ext",
        cfg.movies_tv_ext_scheduler_crontab,
        cfg.disable_movies_tv_ext_scheduler,
        Some("false")
    );
    spider!(
        "nowmetv",
        cfg.nowmetv_scheduler_crontab,
        cfg.disable_nowmetv_scheduler
    );
    spider!(
        "nowsports",
        cfg.nowsports_scheduler_crontab,
        cfg.disable_nowsports_scheduler
    );
    spider!(
        "tamilultra",
        cfg.tamilultra_scheduler_crontab,
        cfg.disable_tamilultra_scheduler
    );
    spider!(
        "sport_video",
        cfg.sport_video_scheduler_crontab,
        cfg.disable_sport_video_scheduler,
        Some("false")
    );
    spider!(
        "dlhd",
        cfg.dlhd_scheduler_crontab,
        cfg.disable_dlhd_scheduler
    );
    spider!(
        "arab_torrents",
        cfg.arab_torrents_scheduler_crontab,
        cfg.disable_arab_torrents_scheduler
    );
    spider!(
        "x1337",
        cfg.x1337_scheduler_crontab,
        cfg.disable_x1337_scheduler,
        Some("false")
    );
    spider!(
        "thepiratebay",
        cfg.thepiratebay_scheduler_crontab,
        cfg.disable_thepiratebay_scheduler,
        Some("false")
    );
    spider!(
        "rutor",
        cfg.rutor_scheduler_crontab,
        cfg.disable_rutor_scheduler,
        Some("false")
    );
    spider!(
        "limetorrents",
        cfg.limetorrents_scheduler_crontab,
        cfg.disable_limetorrents_scheduler,
        Some("false")
    );
    spider!(
        "yts",
        cfg.yts_scheduler_crontab,
        cfg.disable_yts_scheduler,
        Some("false")
    );
    spider!(
        "bt4g",
        cfg.bt4g_scheduler_crontab,
        cfg.disable_bt4g_scheduler,
        Some("false")
    );
    spider!(
        "nyaa",
        cfg.nyaa_scheduler_crontab,
        cfg.disable_nyaa_scheduler,
        Some("false")
    );
    spider!(
        "animetosho",
        cfg.animetosho_scheduler_crontab,
        cfg.disable_animetosho_scheduler,
        Some("false")
    );
    spider!(
        "subsplease",
        cfg.subsplease_scheduler_crontab,
        cfg.disable_subsplease_scheduler,
        Some("false")
    );
    spider!(
        "animepahe",
        cfg.animepahe_scheduler_crontab,
        cfg.disable_animepahe_scheduler,
        Some("false")
    );
    spider!(
        "bt52",
        cfg.bt52_scheduler_crontab,
        cfg.disable_bt52_scheduler,
        Some("false")
    );
    spider!(
        "uindex",
        cfg.uindex_scheduler_crontab,
        cfg.disable_uindex_scheduler,
        Some("false")
    );
    spider!(
        "eztv_rss",
        cfg.eztv_rss_scheduler_crontab,
        cfg.disable_eztv_rss_scheduler
    );

    // ── Feed scrapers & simple tasks ─────────────────────────────────────────
    //
    // Most tasks accept **kwargs and Python passes crontab_expression in kwargs,
    // so we mirror that. Exceptions:
    //   - run_all_integration_syncs: no **kwargs in signature, must not get extra kwargs.

    // with_crontab: passes crontab_expression in kwargs (matches Python)
    macro_rules! simple_job {
        ($name:literal, $actor:literal, $queue:literal, $crontab:expr, $disabled:expr) => {
            if !$disabled {
                let crontab = $crontab.clone();
                let ct_kw = crontab.clone();
                spawn_job($name, crontab, Arc::clone(&state), move |s| {
                    let ct = ct_kw.clone();
                    async move { enqueue(&s, $actor, $queue, json!({"crontab_expression": ct})).await; }
                });
            }
        };
        // always-enabled variant (no disable flag)
        ($name:literal, $actor:literal, $queue:literal, $crontab:expr) => {
            {
                let crontab = $crontab.clone();
                let ct_kw = crontab.clone();
                spawn_job($name, crontab, Arc::clone(&state), move |s| {
                    let ct = ct_kw.clone();
                    async move { enqueue(&s, $actor, $queue, json!({"crontab_expression": ct})).await; }
                });
            }
        };
    }

    // no_kwargs: does NOT pass crontab_expression (for tasks with no **kwargs)
    macro_rules! simple_job_no_kwargs {
        ($name:literal, $actor:literal, $queue:literal, $crontab:expr, $disabled:expr) => {
            if !$disabled {
                let crontab = $crontab.clone();
                spawn_job($name, crontab, Arc::clone(&state), |s| async move {
                    enqueue(&s, $actor, $queue, json!({})).await;
                });
            }
        };
    }

    simple_job!(
        "prowlarr_feed_scraper",
        "run_prowlarr_feed_scraper",
        "scrapy",
        cfg.prowlarr_feed_scraper_crontab,
        cfg.disable_prowlarr_feed_scraper
    );
    simple_job!(
        "jackett_feed_scraper",
        "run_jackett_feed_scraper",
        "scrapy",
        cfg.jackett_feed_scraper_crontab,
        cfg.disable_jackett_feed_scraper
    );
    simple_job!(
        "rss_feed_scraper",
        "run_rss_feed_scraper",
        "scrapy",
        cfg.rss_feed_scraper_crontab,
        cfg.disable_rss_feed_scraper
    );

    // ── Background scrapers ───────────────────────────────────────────────────

    // DMM hashlist scraper reuses the same disable flag as the live scraper
    if !cfg.disable_dmm_hashlist_scraper {
        simple_job!(
            "dmm_hashlist_scraper",
            "run_dmm_hashlist_scraper",
            "scrapy",
            cfg.dmm_hashlist_scraper_crontab
        );
    }

    simple_job!(
        "youtube_background_scraper",
        "run_youtube_background_scraper",
        "scrapy",
        cfg.youtube_background_scraper_crontab,
        cfg.disable_youtube_background_scraper
    );
    simple_job!(
        "acestream_background_scraper",
        "run_acestream_background_scraper",
        "scrapy",
        cfg.acestream_background_scraper_crontab,
        cfg.disable_acestream_background_scraper
    );
    simple_job!(
        "telegram_background_scraper",
        "run_telegram_background_scraper",
        "scrapy",
        cfg.telegram_background_scraper_crontab,
        cfg.disable_telegram_background_scraper
    );

    // ── Maintenance tasks ─────────────────────────────────────────────────────

    simple_job!(
        "validate_tv_streams_in_db",
        "validate_tv_streams_in_db",
        "default",
        cfg.validate_tv_streams_in_db_crontab,
        cfg.disable_validate_tv_streams_in_db
    );
    simple_job!(
        "update_torrent_seeders",
        "update_torrent_seeders",
        "default",
        cfg.update_seeders_crontab,
        cfg.disable_update_seeders
    );
    simple_job!(
        "cleanup_expired_scraper_task",
        "cleanup_expired_scraper_task",
        "priority",
        cfg.cleanup_expired_scraper_task_crontab
    );
    simple_job!(
        "cleanup_expired_cache",
        "cleanup_expired_cache",
        "priority",
        cfg.cleanup_expired_cache_task_crontab
    );
    simple_job!(
        "run_background_search",
        "run_background_search",
        "default",
        cfg.background_search_crontab
    );

    // run_all_integration_syncs has no **kwargs — must not receive extra keyword args
    simple_job_no_kwargs!(
        "run_all_integration_syncs",
        "run_all_integration_syncs",
        "default",
        cfg.integration_sync_crontab,
        cfg.disable_integration_sync_scheduler
    );

    // Discover pre-warm: only when discover is enabled and a TMDB key is set
    if cfg.discover_enabled && cfg.tmdb_api_key.is_some() {
        simple_job!(
            "discover_prewarm",
            "run_discover_prewarm",
            "default",
            "0 4 * * *".to_string()
        );
    }

    info!("scheduler: all jobs registered");
}
